activesupport/lib/active_support/duration.rb

```diff
@@ -89,21 +89,82 @@ def ago(time = ::Time.current)
     def inspect #:nodoc:
       parts.
         reduce(::Hash.new(0)) { |h,(l,r)| h[l] += r; h }.
-        sort_by {|unit,  _ | [:years, :months, :days, :minutes, :seconds].index(unit)}.
+        sort_by {|unit,  _ | [:years, :months, :weeks, :days, :hours, :minutes, :seconds].index(unit)}.
         map     {|unit, val| "#{val} #{val == 1 ? unit.to_s.chop : unit.to_s}"}.
         to_sentence(locale: ::I18n.default_locale)
     end
 
+    # JSON representation should be a string formatted as ISO 8601 Duration, as recommended by Google JSON Style Guide:
+    # https://google-styleguide.googlecode.com/svn/trunk/jsoncstyleguide.xml?showone=Time_Duration_Property_Values#Time_Duration_Property_Values
     def as_json(options = nil) #:nodoc:
-      to_i
+      iso8601(options && options[:precision])
     end
 
     def respond_to_missing?(method, include_private=false) #:nodoc:
       @value.respond_to?(method, include_private)
     end
 
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, it will raise +ActiveSupport::Duration::ISO8601DurationParser::ParsingError+.
+    def self.parse!(iso8601duration)
+      require 'active_support/duration/iso8601_duration_parser' unless defined?(ISO8601DurationParser)
+      parts = ISO8601DurationParser.new(iso8601duration).parts
+      time  = ::Time.now
+      new(time.advance(parts) - time, parts)
+    end
+
+    # Creates a new Duration from string formatted according to ISO 8601 Duration.
+    #
+    # See http://en.wikipedia.org/wiki/ISO_8601#Durations
+    # This method allows negative parts to be present in pattern.
+    # If invalid string is provided, nil will be returned.
+    def self.parse(iso8601duration)
+      parse!(iso8601duration)
+    rescue ISO8601DurationParser::ParsingError
+      nil
+    end
+
+    # Build ISO 8601 Duration string for this duration.
+    # The +precision+ parameter can be used to limit seconds' precision of duration.
+    def iso8601(precision=nil)
```

Comment:
Well, I just [have told to extract this](https://github.com/rails/rails/pull/16917#discussion_r17522955) to separate class (as parsing takes many kines of code due to various validations and so on). For me it looks natural for class to be able to (de)serialize its objects from/to standard format.
I will think about that and probably I will create a new class, something like `ActiveSupport::Duration::ISO8601Serializer` to handle only export of durations to strings in standartized manner :-) (currently, `ISO8601Parser` does not know anything about `Duration`, it just creates compatible data structure to be fed to `Duration`'s initializer)